### DISCUSSION TITLE
Example for submitting binary data from golang to WebUI JavaScript?

### DISCUSSION THREAD

**Discussion by Kukulkano** [webui-dev/webui#160] [General]
Hi,
is there some example available that shows how binary data is submitted to JavaScript using go backend? Specifically, I need a file opened by golang to be submitted as UInt8Array or Array of bytes. I want to avoid any conversation to string or json.

**Comment by AlbertShown** [comment]
I have already taught about this, but the inverse, sends binary data from UI to backend. Thankfully, the original developer of webui set the communication in binary mode since the beginning, so it's easy for us to add a new command in `_webui_window_receive()` to accept a specific number of bytes. And I'm sure we can do the same to send it back to the UI.

However, for now, this should be a new requested feature.

**Reply by Kukulkano** [reply]
Thanks Albert. I'm surprised because the feature list in the documentation main page already mentions "_Fast binary communication protocol between WebUI and the browser (Instead of JSON)_". I assumed there is already some binary interface for in- and outgoing data? So it would need a new function to WebUI for this first? Or for the golang wrapper? 

I came to this question because we need to transfer binary file data with up to 100MB from golang to the WASM component in WebUI. The WASM component can handle this very fast. But to convert data, especially to string or json back and forth, would be a bottleneck for sure.

Together with the missing stable branch, I start to doubt that WebUI is already on a level that is usable for a commercial project like ours. What do you think?

**Reply by AlbertShown** [reply]
> Fast binary communication protocol

The function `_webui_window_receive()` reads the first 4 bytes to know what to do, so we can easily patch it to determine the size of data in bytes, then send it to your Go backend or to your WASM. Anyway, it's an interesting feature to have.

https://github.com/webui-dev/webui/issues/161

**Reply by AlbertShown** [reply]
> Instead of JSON

If webui is designed in text mode, adding this binary feature now will be a hard task. But right now, it's a 1 hour work only.

**Comment by hassandraga** [comment]
Implemented in C, soon will be added to the Golang wrapper.

**C**
```c
unsigned char buffer[] = {0x41, 0x42, 0x43};
webui_send_raw(window, "my_js_function", buffer, 3);
```

**JS**
```js
function my_js_function(buf){
    // buf is a Uint8Array[0x41, 0x42, 0x43]
}
```

**Reply by Kukulkano** [reply]
Hi. Is there a timeline when it is added to the golang wrapper? I plan to make a POC using golang as backend and need to submit binary array data.
